### ISSUE TITLE
Reproducibility on real dataset

### ISSUE THREAD

**Issue by yxie20** [facebookresearch/NSVF#25]
Hello! Thank you for open-sourcing this amazing work.

I have attempted to reproduce your research results on real, inward facing dataset and struggled to achieve photo-realistic results. After a thorough process of elimination, I make an educated guess that the problem is with NSVF's robustness to slightly perturbated camera poses. Running on the same colmap camera poses, we observed that the original NeRF model is significantly more robust to camera pose inaccuracies.

To reproduce: Run colmap on synthetic Lego to get non-ground-truth camera poses. Train model with original configurations.

Here is an example:
Run colmap on synthetic Lego dataset, with exhaustive matching.
<img src="https://user-images.githubusercontent.com/40676957/98009278-5c578580-1dc3-11eb-8b26-f566b119f742.png" height="200">

Now we compare training results on three sets of camera poses: 1) ground truth, 2) ground truth + additive gaussian noise (std=0.01), 3) colmap camera poses in NSVF's coordinate frame convention. We can see that colmap camera poses does not produce desirable result.

<img src="https://user-images.githubusercontent.com/40676957/98010093-4eeecb00-1dc4-11eb-80ff-7c0f8a97eddf.gif" height="200">

<img src="https://user-images.githubusercontent.com/40676957/98010095-4f876180-1dc4-11eb-8d6f-5e65b5fabe41.gif" height="200">

<img src="https://user-images.githubusercontent.com/40676957/98010091-4e563480-1dc4-11eb-8c2c-0908f17833d1.gif" height="200">

Here is a snapshot of the loss function (pink=ground truth poses, blue=colmap). Note that training on colmap camera poses cannot converge to the same photorealistic results _regardless of training time_.
<img src="https://user-images.githubusercontent.com/40676957/98010898-3fbc4d00-1dc5-11eb-94ab-955731c8fbd9.png" height="200">

Here is another example on a real captured inward facing dataset:
<img src="https://user-images.githubusercontent.com/40676957/98011258-9d509980-1dc5-11eb-8d60-d0377f126ab2.gif" height="200">
<img src="https://user-images.githubusercontent.com/40676957/98011271-9fb2f380-1dc5-11eb-8454-e8bf6e7d8518.png" height="200">

**Comment by MultiPath** [comment]
Hi, thanks for posting this! It is interesting that NSVF will be more sensitive to camera errors compared to the original NeRF. In my view, they should face the same learning issue as the ray-marching part is the same.
Also we did not observe the same phenomena on the real datasets we have tested compared to the original NeRF.

1) Can you please share the three Lego dataset for us to take a look?
2) In our code, we also implement an option to fix the camera pose errors (while this function is sometimes not stable)
by adding ``--trainable-extrinsics`` in the training arguments. 
3) Decreasing the learning rate might also be helpful. Too speed up learning, we usually used ``lr=0.001``. You can try training with ``lr=0.0001``.

**Comment by yxie20** [comment]
Thank you for such prompt reply! I've uploaded the data here: https://github.com/yxie20/lego_nsvf. To use it, simply copy the following: `pose`, `bbox.txt`, `intrinsics.txt` to the root folder. The original NSVF folder structure is followed here.

I'll look into your comment 2 and 3. Thanks!

**Comment by yxie20** [comment]
A note on parsing colmap camera K Rt: I used the the handy script provided by the NeRF research group (https://github.com/Fyusion/LLFF/tree/master/llff/poses). The negation (compared to the NeRF/LLFF convention) of the 2nd and 3rd column of camera extrinsics is accounted for.

**Comment by MultiPath** [comment]
> Thank you for such prompt reply! I've uploaded the data here: https://github.com/yxie20/lego_nsvf. To use it, simply copy the following: `pose`, `bbox.txt`, `intrinsics.txt` to the root folder. The original NSVF folder structure is followed here.
> 
> I'll look into your comment 2 and 3. Thanks!

Can you also share some comparison with NeRF where you observe the clear difference?
Thanks!

**Comment by MultiPath** [comment]
> A note on parsing colmap camera K Rt: I used the the handy script provided by the NeRF research group (https://github.com/Fyusion/LLFF/tree/master/llff/poses). The negation (compared to the NeRF/LLFF convention) of the 2nd and 3rd column of camera extrinsics is accounted for.

Ok. We are using OpenCV style coordinates for extrinsics. Do you mean we need to manually change the code to run your Colmap data?

**Comment by yxie20** [comment]
> > A note on parsing colmap camera K Rt: I used the the handy script provided by the NeRF research group (https://github.com/Fyusion/LLFF/tree/master/llff/poses). The negation (compared to the NeRF/LLFF convention) of the 2nd and 3rd column of camera extrinsics is accounted for.
> 
> Ok. We are using OpenCV style coordinates for extrinsics. Do you mean we need to manually change the code to run your Colmap data?

No, not at all! Everything is accounted for and you only need to copy, past and run!

**Comment by yxie20** [comment]
> > Thank you for such prompt reply! I've uploaded the data here: https://github.com/yxie20/lego_nsvf. To use it, simply copy the following: `pose`, `bbox.txt`, `intrinsics.txt` to the root folder. The original NSVF folder structure is followed here.
> > I'll look into your comment 2 and 3. Thanks!
> 
> Can you also share some comparison with NeRF where you observe the clear difference?
> Thanks!

#Here's a NeRF training for 40 minutes or 5k steps, with colmap poses on Lego:
<img src="https://user-images.githubusercontent.com/40676957/98016920-8d888380-1dcc-11eb-87c0-575a31706d54.gif" height="200">

**Comment by yxie20** [comment]
To help us better pinpoint the problem, I added my pose parsing code to https://github.com/yxie20/lego_nsvf. This code takes in colmap .bin files and output `pose/`, `bbox.txt` and `intrinsics.txt` in NSVF standard. The initial voxel size (side length) is automatically set to 1/8 of the total volume side length.

**Comment by yxie20** [comment]
Here I provide some results with --trainable-extrinsics enabled. The result is still far from that achieved by ground truth poses.
<img src="https://user-images.githubusercontent.com/40676957/98195153-0b44b000-1eef-11eb-8242-17076379813c.png" width=100px>
Here is the loss curve comparison (purple is gt poses, blue is colmap+trainable-extrinsics)
<img src="https://user-images.githubusercontent.com/40676957/98195240-3b8c4e80-1eef-11eb-97ae-90adbb6e5b17.png" width=100px>

**Comment by kwea123** [comment]
Telling from the depth map, it seems that it doesn't learn the correct structure either...
Haven't tried NSVF yet, but I have a doubt on your NeRF result, because in my experiments, it works pretty well on real world objects for which the poses are also estimated by colmap. 
From your very first picture already, I doubt that the poses are not correct at all, because the gt poses should be distributed **uniformly** on the upper hemisphere if I remember correctly; your reconstructed poses seem very irregular. I suspect that it is due to the white background which makes the pose reconstruction difficult. I would suggest that you align the gt poses and the reconstructed poses to see how much they differ, before digging into either NeRF or NSVF.

**Comment by yxie20** [comment]
> Telling from the depth map, it seems that it doesn't learn the correct structure either...
> Haven't tried NSVF yet, but I have a doubt on your NeRF result, because in my experiments, it works pretty well on real world objects for which the poses are also estimated by colmap.
> From your very first picture already, I doubt that the reconstructed poses are not correct at all, because the gt poses are distributed rather **uniformly** on the upper hemisphere if I remember correctly like in [Fig. 1](https://arxiv.org/pdf/2003.08934.pdf) ; your reconstructed poses seem very irregular. I suspect that it is due to the white background which makes the pose reconstruction difficult. I would suggest that you align the gt poses and the reconstructed poses to see how much they differ, before digging into either NeRF or NSVF.

Thanks for the comment. If you do decide to run NSVF, I look forward to hearing how it goes! Note that the same colmap poses were used in NeRF and NSVF, that's the key of this thread here.

**Comment by MultiPath** [comment]
Hi I haven't had time looking into it yet. I will take a look at this issue this weekend

**Comment by tau-yihouxiang** [comment]
@yxie20 Have you solve this issue? I decided to run colmap on real-world dataset by myself.

**Comment by yxie20** [comment]
No... Still waiting on the author @MultiPath to provide some insights.

**Comment by MultiPath** [comment]
Sorry about that. I was busy with other papers recently. will debug this for a bit

**Comment by kondela** [comment]
I have similar experience as @yxie20.

NSVF - default configuration, 62.5k steps
![image (2)](https://user-images.githubusercontent.com/34348494/105222494-d72a7400-5b5a-11eb-8941-5ecda220a8cd.png)

NSVF - enabled `trainable_extrinsics` and lowered learning rate to 0.0001 as suggested by @MultiPath . 25k steps then OOM
![image (1)](https://user-images.githubusercontent.com/34348494/105222343-ac402000-5b5a-11eb-9c33-7174a35848a4.png)

nerf (@kwea123 implementation) - original, validation image, and depth map.
![image (3)](https://user-images.githubusercontent.com/34348494/105222626-0b9e3000-5b5b-11eb-92c3-b5ebe0be2515.png)

I used same images and poses from COLMAP for all experiments.

**Comment by MultiPath** [comment]
Hi @kondela can you also share your dataset? if possible

**Comment by kondela** [comment]
Hey @MultiPath, sure! You can download it from my [google drive](https://drive.google.com/file/d/1o0WHu9xMWBb6LW26lgKmtFKxzfZEmPtW/view?usp=sharing). I used first 142 images for training and left last 2 for validation.

**Comment by MultiPath** [comment]
Hi @kondela, just for double check.. Are you sure the pose files and the images are aligned? In my code, I sorted the poses and images separately by string order.
![image](https://user-images.githubusercontent.com/5780274/105645045-304e1c80-5e67-11eb-85b2-34dc75b0d469.png)

**Comment by kondela** [comment]
Hey @MultiPath, you might be right that poses and images are not aligned properly. 

I am using @yxie20 script (https://github.com/yxie20/lego_nsvf/blob/master/poses/pose_utils.py) to generate poses, intrinsics and bounding box as well, and it seems that there's assumption that `cameras.bin` from COLMAP contains sorted poses based on image file names (which it does not). This would explain why both of us have poor results on real datasets.

In original LLFF they use `np.argsort()` to account for it (see variable `perm` in https://github.com/Fyusion/LLFF/blob/c6e27b1ee59cb18f054ccb0f87a90214dbe70482/llff/poses/pose_utils.py#L33). I tried to fix it and ran training using default config, however, it seems that results are still poor. 
![image](https://user-images.githubusercontent.com/34348494/105822535-9e185680-5fd5-11eb-8e03-0d94f872a725.png)
 
You can download the new [pose.zip here](https://github.com/facebookresearch/NSVF/files/5872247/pose.zip).

I will take a look at it again, there might be another bug that causes improper alignment between images and poses or something similar.

 Meanwhile, if you have time it would be awesome if you could provide your own script to convert COLMAP's model files (`cameras.bin`, `images.bin`, `points3D.bin`) to format that NSVF expects, i.e. `bbox.txt`, `intrinsics.txt` and poses folder.

**Comment by dedoogong** [comment]
I also met this problem. I think, to reproduce the original result with other datasets, I would like to ask author to give the code for that( camera.bin/image.bin to bbox.txt, intrinsic.txt)

Please help us! thank you~!
